packages/server-boilerplate/src/routes/Route.ts

```diff
@@ -1,36 +1,37 @@
-/*
+/**
  * Tupaia
- * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
- *
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
  */
-import { NextFunction, Response } from 'express';
-import { respond, UnauthenticatedError } from '@tupaia/utils';
-
-import { TupaiaRequest, TupaiaResponseBody, SessionCookie } from '../types';
-import { SessionModel, SessionType } from '../models/Session';
-
-export abstract class Route {
-  req: TupaiaRequest;
-
-  res: Response;
-
-  next: NextFunction;
-
-  sessionModel: SessionModel;
-
-  sessionCookie?: SessionCookie;
-
-  constructor(req: TupaiaRequest, res: Response, next: NextFunction) {
+import { Request, NextFunction, Response } from 'express';
+import { respond } from '@tupaia/utils';
+
+// Infers type arguments from request type
+/* eslint-disable @typescript-eslint/no-explicit-any */
+export type Params<Type> = Type extends Request<infer P, any, any, any> ? P : never;
+export type ResBody<Type> = Type extends Request<any, infer sBody, any, any> ? sBody : never;
+export type ReqBody<Type> = Type extends Request<any, any, infer qBody, any> ? qBody : never;
+export type Query<Type> = Type extends Request<any, any, any, infer Q> ? Q : never;
+export type MatchingRequest<Req> = Request<Params<Req>, ResBody<Req>, ReqBody<Req>, Query<Req>>;
+export type MatchingResponse<Req> = Response<ResBody<Req>>;
```

Comment:
This is the generics hell that Express has put me through. I've tried my best to find a way to define `Route` as being able to take a type that extends `Request` but shares `Request`'s generic parameters. It feels odd to me that typescript doesn't have a good syntax to support this... Atleast all this complexity is now encapsulated in the `server-boilerplate package` 👍

Reply:
Yeap at least it's good that you moved this to a central location! Even as is, the verbosity is worth the extra type safety